quic/s2n-quic-core/src/crypto/tls/offload.rs

```diff
@@ -0,0 +1,567 @@
+                    let resp = context.receive_handshake(max_len);
+                    if resp.is_some() {
+                        // We need to wake up the s2n-quic endpoint after providing
+                        // handshake packets to the TLS provider as there may now be
+                        // handshake data that needs to be sent in response.
+                        context.waker().wake_by_ref();
+                    }
+                    let _ = sender.send(resp);
+                }
+                Request::CanSendHandshake(sender) => {
+                    let _ = sender.send(context.can_send_handshake());
+                }
+                Request::CanSendApplication(sender) => {
+                    let _ = sender.send(context.can_send_application());
+                }
+                Request::SendApplication(bytes) => {
+                    context.send_application(bytes);
+                }
+                Request::SendHandshake(bytes) => {
+                    context.send_handshake(bytes);
+                }
+                Request::SendInitial(bytes) => context.send_initial(bytes),
+                Request::KeyExchangeGroup(named_group) => {
+                    context.on_key_exchange_group(named_group)?;
+                }
+                Request::TlsContext(ctx) => context.on_tls_context(ctx),
+            }
+        }
+        Poll::Pending
+    }
+}
+
+impl<S: tls::Session> CryptoSuite for OffloadSession<S> {
+    type HandshakeKey = <S as CryptoSuite>::HandshakeKey;
+    type HandshakeHeaderKey = <S as CryptoSuite>::HandshakeHeaderKey;
+    type InitialKey = <S as CryptoSuite>::InitialKey;
+    type InitialHeaderKey = <S as CryptoSuite>::InitialHeaderKey;
+    type ZeroRttKey = <S as CryptoSuite>::ZeroRttKey;
+    type ZeroRttHeaderKey = <S as CryptoSuite>::ZeroRttHeaderKey;
+    type OneRttKey = <S as CryptoSuite>::OneRttKey;
+    type OneRttHeaderKey = <S as CryptoSuite>::OneRttHeaderKey;
+    type RetryKey = <S as CryptoSuite>::RetryKey;
+}
+
+struct AsyncRequest<T> {
+    rx: Option<Receiver<T>>,
+}
+
+impl<T> AsyncRequest<T> {
+    fn empty() -> Self {
+        AsyncRequest { rx: None }
+    }
+
+    fn poll_request(
+        &mut self,
+        cx: &mut core::task::Context<'_>,
+        issue: impl FnOnce(Sender<T>),
+    ) -> Poll<T> {
+        loop {
+            if let Some(mut receiver) = self.rx.as_mut() {
+                match Pin::new(&mut receiver).poll(cx) {
+                    Poll::Ready(Ok(value)) => {
+                        receiver.close();
+                        self.rx = None;
+                        return Poll::Ready(value);
+                    }
+                    Poll::Ready(Err(_)) => {
+                        // treat cancellation as reason to ask again.
+                        // FIXME: this probably means that the parent thread is no longer interested
+                        // in this connection and we should instead tear it down.
```

Comment:
can this be fixed now or is there more investigation needed? If it can't be fixed now, would be good to link an issue